### PR TITLE
Create lazy read-only database sessions

### PR DESCRIPTION
**Pull request by gmethvin**
In many cases we'll make a database call like

``` scala
db.readOnly { implicit s => userRepo.get(id) }
```

which will most of the time get the value from memcache and never use the database connection we just opened. This makes the database session lazy so we don't initialize it until we actually use it.

### CODE REVIEW

**Comment by andrewconner** [comment]
Great :)

**Comment by eishay** [comment]
Brilliant !
I wonder if there's a way to test it. I would be concerned that we'll instantiate the session even though we don't really need it. E.g. the userRepo will use the memcache for the get(id) method but for some reason the session will get instantiated anyway by the UserRepo class.

File: shoebox/app/com/keepit/common/db/slick/DBSession.scala [hunk 1]
```diff
@@ -34,9 +36,9 @@ object DBSession {
       rsHoldability: ResultSetHoldability = resultSetHoldability) = throw new UnsupportedOperationException
   }
 
-  abstract class RSession(roSession: Session) extends SessionWrapper(roSession)
-  class ROSession(roSession: Session) extends RSession(roSession)
-  class RWSession(rwSession: Session) extends RSession(rwSession)
+  abstract class RSession(roSession: => Session) extends SessionWrapper(roSession)
+  class ROSession(roSession: => Session) extends RSession(roSession)
+  class RWSession(rwSession: => Session) extends RSession(rwSession)
```
Comment (2is10) [inline shoebox/app/com/keepit/common/db/slick/DBSession.scala]: why update `RWSession` to call by name if not also updating any callers? leaving it as it was might help `ROSession` draw the special attention from readers that it deserves.

**Comment by 2is10** [comment]
I'm in favor of this idea if done safely and correctly. At a glance this looks okay but I think Eishay knows this code best.

**Comment by 2is10** [comment]
Good point Eishay. Do we need to change all the repos to take an implicit `=> Session` instead of `Session`?

**Comment by andrewconner** [comment]
How did you ensure the session isn't being created?

We could test it by providing a mock `Session` that throws an exception if used. That doesn't test `readOnly` directly, though.

File: shoebox/app/com/keepit/common/db/slick/DBSession.scala (hunk 1 above)
Comment (gmethvin) [inline shoebox/app/com/keepit/common/db/slick/DBSession.scala]: Good point. I was originally trying to make RWSession lazy too but then I realized we didn't really need it there.

**Comment by gmethvin** [comment]
@2is10 We don't need to change the repos, since RSession wraps the original session implementation.

I tested by printing the stack trace using an exception in dev mode when I initialized the session, and made sure it wasn't called when I made requests that called repo methods that used memcache.

I think there should probably be a test though. I was thinking of something along the lines of Andrew's suggestion.

**Comment by gmethvin** [comment]
To test this I factored out a SlickSessionProvider which lets us monitor when we create sessions. In tests I count the number of database sessions we initialize so we can make sure we're going to cache.

**Comment by 2is10** [comment]
Cool. Where's the test?

**Comment by 2is10** [comment]
Nevermind... I was looking at the wrong commit.

**Pull request merged**